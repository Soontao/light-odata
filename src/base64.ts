/* eslint-disable max-len */
/* eslint-disable prefer-const */
// Array of bytes to Base64 string decoding
function b64ToUint6(nChr) {
  return nChr > 64 && nChr < 91 ?
    nChr - 65
    : nChr > 96 && nChr < 123 ?
      nChr - 71
      : nChr > 47 && nChr < 58 ?
        nChr + 4
        : nChr === 43 ?
          62
          : nChr === 47 ?
            63
            :
            0;
}

function base64DecToArr(sBase64: string, nBlocksSize?: number) {
  const sB64Enc = sBase64.replace(/[^A-Za-z0-9+/]/g, "");
  const nInLen = sB64Enc.length;
  const nOutLen = nBlocksSize ? Math.ceil((nInLen * 3 + 1 >> 2) / nBlocksSize) * nBlocksSize : nInLen * 3 + 1 >> 2;
  const taBytes = new Uint8Array(nOutLen);

  let nMod3;
  let nMod4;
  let nUint24 = 0;
  let nOutIdx = 0;
  for (let nInIdx = 0; nInIdx < nInLen; nInIdx++) {
    nMod4 = nInIdx & 3;
    nUint24 |= b64ToUint6(sB64Enc.charCodeAt(nInIdx)) << 6 * (3 - nMod4);
    if (nMod4 === 3 || nInLen - nInIdx === 1) {
      nMod3 = 0;
      while (nMod3 < 3 && nOutIdx < nOutLen) {
        taBytes[nOutIdx] = nUint24 >>> (16 >>> nMod3 & 24) & 255;
        nMod3++;
        nOutIdx++;
      }
      nUint24 = 0;

    }
  }

  return taBytes;
}

/* Base64 string to array encoding */
function uint6ToB64(nUint6: number) {
  return nUint6 < 26 ?
    nUint6 + 65
    : nUint6 < 52 ?
      nUint6 + 71
      : nUint6 < 62 ?
        nUint6 - 4
        : nUint6 === 62 ?
          43
          : nUint6 === 63 ?
            47
            :
            65;
}

function base64EncArr(aBytes) {
  let nMod3 = 2;
  let sB64Enc = "";

  const nLen = aBytes.length;
  let nUint24 = 0;
  for (let nIdx = 0; nIdx < nLen; nIdx++) {
    nMod3 = nIdx % 3;
    if (nIdx > 0 && (nIdx * 4 / 3) % 76 === 0) {
      sB64Enc += "\r\n";
    }

    nUint24 |= aBytes[nIdx] << (16 >>> nMod3 & 24);
    if (nMod3 === 2 || aBytes.length - nIdx === 1) {
      sB64Enc += String.fromCodePoint(uint6ToB64(nUint24 >>> 18 & 63), uint6ToB64(nUint24 >>> 12 & 63), uint6ToB64(nUint24 >>> 6 & 63), uint6ToB64(nUint24 & 63));
      nUint24 = 0;
    }
  }
  return sB64Enc.substring(0, sB64Enc.length - 2 + nMod3) + (nMod3 === 2 ? "" : nMod3 === 1 ? "=" : "==");
}

/* UTF-8 array to JS string and vice versa */

function utf8ArrToStr(aBytes) {
  let sView = "";
  let nPart;
  const nLen = aBytes.length;
  for (let nIdx = 0; nIdx < nLen; nIdx++) {
    nPart = aBytes[nIdx];
    sView += String.fromCodePoint(
      nPart > 251 && nPart < 254 && nIdx + 5 < nLen ? /* six bytes */
        /* (nPart - 252 << 30) may be not so safe in ECMAScript! So…: */
        (nPart - 252) * 1073741824 + (aBytes[++nIdx] - 128 << 24) + (aBytes[++nIdx] - 128 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
        : nPart > 247 && nPart < 252 && nIdx + 4 < nLen ? /* five bytes */
          (nPart - 248 << 24) + (aBytes[++nIdx] - 128 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
          : nPart > 239 && nPart < 248 && nIdx + 3 < nLen ? /* four bytes */
            (nPart - 240 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
            : nPart > 223 && nPart < 240 && nIdx + 2 < nLen ? /* three bytes */
              (nPart - 224 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
              : nPart > 191 && nPart < 224 && nIdx + 1 < nLen ? /* two bytes */
                (nPart - 192 << 6) + aBytes[++nIdx] - 128
                : /* nPart < 127 ? */ /* one byte */
                nPart
    );
  }
  return sView;
}

function strToUTF8Arr(sDOMStr: string): Uint8Array {
  let aBytes;
  let nChr;
  const nStrLen = sDOMStr.length;
  let nArrLen = 0;

  /* mapping… */
  for (let nMapIdx = 0; nMapIdx < nStrLen; nMapIdx++) {
    nChr = sDOMStr.codePointAt(nMapIdx);

    if (nChr > 65536) {
      nMapIdx++;
    }

    nArrLen += nChr < 0x80 ? 1 : nChr < 0x800 ? 2 : nChr < 0x10000 ? 3 : nChr < 0x200000 ? 4 : nChr < 0x4000000 ? 5 : 6;
  }

  aBytes = new Uint8Array(nArrLen);

  /* transcription… */
  let nIdx = 0;
  let nChrIdx = 0;
  while (nIdx < nArrLen) {
    nChr = sDOMStr.codePointAt(nChrIdx);
    if (nChr < 128) {
      /* one byte */
      aBytes[nIdx++] = nChr;
    } else if (nChr < 0x800) {
      /* two bytes */
      aBytes[nIdx++] = 192 + (nChr >>> 6);
      aBytes[nIdx++] = 128 + (nChr & 63);
    } else if (nChr < 0x10000) {
      /* three bytes */
      aBytes[nIdx++] = 224 + (nChr >>> 12);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
    } else if (nChr < 0x200000) {
      /* four bytes */
      aBytes[nIdx++] = 240 + (nChr >>> 18);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    } else if (nChr < 0x4000000) {
      /* five bytes */
      aBytes[nIdx++] = 248 + (nChr >>> 24);
      aBytes[nIdx++] = 128 + (nChr >>> 18 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    } else /* if (nChr <= 0x7fffffff) */ {
      /* six bytes */
      aBytes[nIdx++] = 252 + (nChr >>> 30);
      aBytes[nIdx++] = 128 + (nChr >>> 24 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 18 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    }
    nChrIdx++;
  }

  return aBytes;
}

const encodeJs = (input: string) => {
  return base64EncArr(strToUTF8Arr(input));
};

const decodeJs = (input: string) => {
  return utf8ArrToStr(base64DecToArr(input));
};

/**
 * encode string to base64
 *
 * @param input
 */
const encode = function (input = ""): string {
  if (globalThis.Buffer !== undefined) {
    return globalThis.Buffer.from(input, "utf-8").toString("base64");
  }
  return base64EncArr(strToUTF8Arr(input));
};

/**
 * decode base64 to string
 *
 * @param input base64 string
 */
const decode = function (input: string): string {
  if (globalThis.Buffer !== undefined) {
    return globalThis.Buffer.from(input, "base64").toString("utf-8");
  }
  return utf8ArrToStr(base64DecToArr(input));
};

export { encode, decode, encodeJs, decodeJs };
