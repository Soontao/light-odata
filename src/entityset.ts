// @ts-nocheck
import { ODataServerError } from "./errors";
import { ODataFilter } from "./filter";
import { systemOptions, SystemQueryOptions } from "./params";
import { OData } from "./odata";
import type { DeepPartial, ODataKeyPredicate, PlainODataResponse } from "./types";

/**
 * EntitySet
 */
export class EntitySet<T = any> {

  private _entitySet: string;

  private _client: OData;

  constructor(collection: string, client: OData) {
    this._entitySet = collection;
    this._client = client;
  }

  private _checkError(res: any): void {
    if (res.error) {
      switch (this._client.getVersion()) {
        case "v2":
          throw new ODataServerError(res.error?.message?.value);
        case "v4":
          throw new ODataServerError(res.error?.message);
        default:
          break;
      }
    }
  }

  private _getResultSingle(res: PlainODataResponse) {
    switch (this._client.getVersion()) {
      case "v2":
        // @ts-ignore
        return res.d;
      case "v4":
        // @ts-ignore
        return res;
      default:
        break;
    }
  }

  private _getResults(res: PlainODataResponse) {
    switch (this._client.getVersion()) {
      case "v2":
        // @ts-ignore
        return res.d?.results;
      case "v4":
        // @ts-ignore
        return res?.value;
      default:
        break;
    }
  }

  public newFilter(): ODataFilter<T> {
    return this._client.newFilter();
  }

  /**
   * @deprecated
   * @returns
   */
  public newParam(): SystemQueryOptions<T> {
    return this._client.newParam();
  }

  public newOptions() {
    return this._client.newOptions();
  }

  async retrieve(id: ODataKeyPredicate, systemOptions?: SystemQueryOptions): Promise<T> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "GET",
      id,
      params: systemOptions
    });
    this._checkError(res);
    return this._getResultSingle(res);
  }


  async find(base: DeepPartial<T>): Promise<T[]> {
    const filter = OData.newFilter();

    Object.entries(base).forEach(([key, value]) => {
      if (typeof value == "string") {
        filter.field(key).eqString(value);
      } else {
        filter.field(key).eq(value);
      }
    });

    return this.query(OData.newParam().filter(filter));
  }

  async query(filter: ODataFilter): Promise<T[]>;

  async query(systemOptions: SystemQueryOptions): Promise<T[]>;

  async query(): Promise<T[]>;

  async query(param: any): Promise<any> {
    if (param instanceof ODataFilter) {
      param = systemOptions().filter(param);
    }
    if (param == undefined) {
      param = OData.newParam();
    }
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "GET",
      params: param
    });
    this._checkError(res);
    return this._getResults(res);
  }

  async count(filter?: ODataFilter): Promise<number>;

  async count(filter?: Partial<T>): Promise<number>;

  async count(filter?: any) {

    const params = OData.newParam().top(1);

    // set count flag
    switch (this._client.getVersion()) {
      case "v4":
        params.count(true);
        break;
      case "v2":
        params.inlinecount(true);
        break;
    }

    if (typeof filter === "object") {

      if (filter instanceof ODataFilter) {
        params.filter(filter);
      } else {
        const newFilter = this.newFilter();
        for (const [key, value] of Object.entries(filter)) {
          newFilter.field(key).eq(value);
        }
        params.filter(newFilter);
      }

    }

    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "GET",
      params
    });

    this._checkError(res);
    switch (this._client.getVersion()) {
      case "v2":
        return parseInt(res?.d?.__count);
      case "v4":
        return res["@odata.count"];
    }
  }

  async create(body: DeepPartial<T>): Promise<T>;

  async create(body: any): Promise<T>;

  async create(body: DeepPartial<T>): Promise<T> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "POST",
      entity: body
    });
    this._checkError(res);
    return this._getResultSingle(res);

  }

  /**
   * update instance by `UPDATE` method
   *
   * @param id
   * @param body
   */
  async update(id: ODataKeyPredicate, body: DeepPartial<T>): Promise<void> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "PATCH",
      id,
      entity: body
    });
    this._checkError(res);
  }

  /**
   * delete instance by object keys
   *
   * @param id
   */
  async delete(id: ODataKeyPredicate): Promise<void> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "DELETE",
      id
    });
    this._checkError(res);
  }

  /**
   * perform bounded action
   *
   * @param actionName the action name, remember add namespace for it
   * @param parameters
   */
  async action(actionName: string, id: ODataKeyPredicate, parameters?: any): any {
    const responseBody = await this._client.newRequest({
      collection: this._entitySet,
      method: "POST",
      id,
      parameters,
      actionName
    });
    this._checkError(responseBody);
    return responseBody;
  }

  /**
   * perform bounded function
   *
   * @param functionName
   * @param id
   */
  async function(
    functionName: string,
    id: ODataKeyPredicate,
    parameters?: any,
    systemOptions?: SystemQueryOptions
  ): any {
    const responseBody = await this._client.newRequest({
      collection: this._entitySet,
      method: "GET",
      id,
      params: systemOptions,
      parameters,
      functionName
    });
    this._checkError(responseBody);
    return responseBody;
  }



}
