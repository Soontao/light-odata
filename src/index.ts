import { OData } from "./odata";

export * from "./batch";
export * from "./filter";
export * from "./params";
export * from "./types";
export { EdmV4 } from "./types_v4";
export * from "./util";
export * from "./request";
export { transformation } from "./tranformation";
export { OData };

export default OData;
