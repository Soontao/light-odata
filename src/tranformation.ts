import { ODataFilter } from "./filter";


/**
 * Transformation for `$apply`
 *
 * @experimental
 */
export class Transformation {

  private _aggregate: string;

  private _countSumPercent: {
    count: number;
    valueExpression: string;
    type: "sum" | "count" | "percent";
    direction: "top" | "bottom";
  };

  private _identity: boolean;

  private _concat: Transformation[];

  private _groupBy: {
    properties: Array<string>;
    transformations: Transformation[];
  };

  private _filter: ODataFilter<any>;

  public static newTransformation() {
    return new Transformation();
  }

  private constructor() { }

  /**
   * Aggregate expressions can define an alias using the as keyword,
   * followed by a SimpleIdentifier
   *
   * @see [Transformation aggregate](http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs01/odata-data-aggregation-ext-v4.0-cs01.html#_The_aggregate_Transformation_1)
   * @param expression
   * @experimental
   * @example
   *
   * ```js
   * aggregate('Amount with sum as Total')
   * ```
   * @returns
   */
  public aggregate(expression: string) {
    this._aggregate = expression;
    return this;
  }

  /**
   *
   * @param count The first parameter specifies the number of instances to return in the transformed set. currently only support number
   * @param valueExpression The second parameter specifies the value by which the instances are compared for determining the result set.
   */
  public topcount(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "top",
      type: "count"
    };
    return this;
  }

  public topsum(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "top",
      type: "sum"
    };
    return this;
  }

  public toppercent(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "top",
      type: "percent",
    };
    return this;
  }


  public bottomcount(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "bottom",
      type: "count"
    };
    return this;
  }

  public bottomsum(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "bottom",
      type: "sum"
    };
    return this;
  }

  public bottompercent(count: number, valueExpression: string) {
    this._countSumPercent = {
      count,
      valueExpression,
      direction: "bottom",
      type: "percent",
    };
    return this;
  }

  /**
   * The identity transformation returns its input set.
   *
   * @returns
   */
  public identity() {
    this._identity = true;
    return this;
  }

  /**
   * The concat transformation takes two or more parameters,
   * each of which is a sequence of set transformations.
   *
   * @param transformations
   * @returns
   */
  public concat(...transformations: Array<Transformation>) {
    this._concat = transformations;
    return this;
  }

  /**
   * The filter transformation takes a Boolean expression that could also be passed as a `$filter` system query option to its input set and returns all instances for which this expression evaluates to true.
   * @param filter
   * @returns
   */
  public filter(filter: ODataFilter) {
    this._filter = filter;
    return this;
  }

  /**
   *
    The `groupby` transformation takes one or two parameters and
    1. Splits the initial set into subsets where all instances in a subset have the same values for the grouping properties specified in the first parameter,
    2. Applies set transformations to each subset according to the second parameter, resulting in a new set of potentially different structure and cardinality,
    3. Ensures that the instances in the result set contain all grouping properties with the correct values for the group,
    4. Concatenates the intermediate result sets into one result set.

   * @see [Transformation groupby](http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs01/odata-data-aggregation-ext-v4.0-cs01.html#_Toc378326304)
   * @param properties
   * @param transformations
   * @experimental
   * @returns
   */
  public groupBy(
    properties: Array<string>,
    ...transformations: Array<Transformation>
  ) {

    // TODO: properties can not be empty
    this._groupBy = {
      properties,
      transformations
    };
    return this;
  }

  // TODO: filter, expand, search, isdefined, $crossjoin


  public toString() {
    const expressionText = "";
    if (this._aggregate !== undefined) {
      return `aggregate(${this._aggregate})`;
    }

    if (this._countSumPercent !== undefined) {
      const { direction, type, count, valueExpression } = this._countSumPercent;
      return `${direction}${type}(${count},${valueExpression})`;
    }

    if (this._groupBy !== undefined) {
      const properties = `(${this._groupBy.properties.join(",")})`;
      if (this._groupBy.transformations.length > 0) {
        const transformations = this
          ._groupBy
          .transformations
          .map(t => t.toString())
          .join("/");
        // see Transformation Sequences
        return `groupBy(${properties},${transformations})`;
      }
      return `groupBy(${properties})`;
    }

    if (this._identity === true) {
      return "identity";
    }

    if (this._filter !== undefined) {
      //  $apply is only avaible for odata v4,
      return `filter(${this._filter.toString()})`;
    }

    // TODO: other transformation

    return expressionText;
  }

}


export function transformation() {
  return Transformation.newTransformation();
}
