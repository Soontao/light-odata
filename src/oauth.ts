import { Mutex } from "@newdash/newdash/functional/Mutex";
import { encode } from "./base64";
import { AuthenticationError } from "./errors";
import { SearchParams } from "./util";

const S_AUTHORIZATION = "Authorization";
const S_BEARER = "Bearer";
const S_CLIENT_CREDENTIALS = "client_credentials";
const S_CT_URL_FORM = "application/x-www-form-urlencoded";

export type TokenRetrieveType = "header" | "form";

/**
 * Client Credential OAuth Client
 *
 * Will automatically process the token refresh logic
 *
 * refresh policy: half expire-in
 *
 * @example
 * ```js
 * // please remember share client instance for a same OAuth server
 * const client = new ClientCredentialsOAuthClient(
 *  "http://example/oauth/token",
 *  "client_id",
 *  "client_secret"
 * )
 * const response = await client.fetch("http://protected.by.token/resource", {
 *  method: "GET",
 * })
 * ```
 *
 *
 */
class ClientCredentialsOAuthClient {

  private readonly tokenUrl: string;

  private readonly clientId: string;

  private readonly clientSecret: string;

  private readonly scope: string;

  private readonly mut = new Mutex();

  private token: string;

  private expiresAt = 0;

  private tokenRetrieveType: TokenRetrieveType = "header";


  /**
   * a very simple oauth client credential client
   *
   * @param tokenUrl oauth token url
   * @param clientId oauth client id
   * @param clientSecret oauth client secret
   * @param retrieveType the clientId and clientSecret is put into header or form body
   * @param scope oauth scope
   */
  constructor(
    tokenUrl: string,
    clientId: string,
    clientSecret: string,
    retrieveType: TokenRetrieveType = "header",
    scope?: string,
  ) {

    this.tokenUrl = tokenUrl;
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.tokenRetrieveType = retrieveType;
    this.scope = scope;
  }

  /**
   * fetch token directly
   * @private
   * @returns {Promise<any>}
   */
  private async fetchOAuthResponse(): Promise<any> {
    const params = new SearchParams();

    params.append("grant_type", S_CLIENT_CREDENTIALS);

    if (typeof this.scope === "string") { params.append("scope", this.scope); }

    let response = undefined;

    if (this.tokenRetrieveType === "form") {

      params.append("client_id", this.clientId);
      params.append("client_secret", this.clientSecret);

      response = await fetch(this.tokenUrl, {
        method: "POST",
        body: params.toString(),
        headers: { "Content-Type": S_CT_URL_FORM }
      });

    }
    else {

      // standard way described in rfc6749
      // https://datatracker.ietf.org/doc/html/rfc6749#section-4.4.2
      response = await fetch(this.tokenUrl, {
        method: "POST",
        body: params.toString(),
        headers: {
          [S_AUTHORIZATION]: "Basic " + encode(`${this.clientId}:${this.clientSecret}`),
          "Content-Type": S_CT_URL_FORM
        }
      });

    }


    if (response.status >= 400) {

      if (response.headers.get("Content-Type")?.includes?.("application/json")) {
        const responseBody = await response.json();
        if ("error" in responseBody) {
          throw new AuthenticationError(responseBody.error);
        }
        else {
          throw new AuthenticationError(JSON.stringify(responseBody));
        }
      }
      else {
        throw new Error(await response.text());
      }
    }

    return response.json();
  }

  /**
   * get header pair
   */
  public async getHeader() {
    const token = await this.getToken();
    return {
      [S_AUTHORIZATION]: `${S_BEARER} ${token}`
    };
  }

  /**
   * get current valid oauth token
   *
   * @returns {Promise<string>}
   */
  public async getToken(): Promise<string> {

    // lock avoid multi fetch token requests in same time
    return this.mut.use(async () => {

      try {

        const currentTime = new Date().getTime();

        // retrive new token
        if (currentTime >= this.expiresAt) {
          const { access_token: accessToken, expires_in: expiresIn } = await this.fetchOAuthResponse();
          this.token = accessToken;
          // half expire
          this.expiresAt = currentTime + ((expiresIn / 2) * 1000);
        }

      } catch (error) {

        console?.error?.(`oauth client retrieve token failed: ${error?.message ?? error}`,);

        // rethrow error
        throw error;

      }

      return this.token;

    });


  }

  /**
   * send requests with oauth token
   *
   */
  public async fetch(...args: Parameters<typeof fetch>) {
    // refresh token if required
    const token = await this.getToken();

    if (token === undefined) {
      throw new TypeError("un-expected token is undefined");
    }

    const url = args[0];
    let config = args[1];

    if (config === undefined) {
      config = {};
    }
    // append jwt token
    config.headers = {
      ...(config.headers || {}),
      [S_AUTHORIZATION]: `${S_BEARER} ${token}`
    };

    return fetch(url, config);
  }

}

export { ClientCredentialsOAuthClient };
