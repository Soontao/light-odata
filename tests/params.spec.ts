import { OData, ODataDateTime, ODataDateTimeOffset, ODataParam, ODataQueryParam, RawString, SystemQueryOptions } from "../src";
import { transformation, Transformation } from "../src/tranformation";

describe('ODataParams Test', () => {

  test('ODataQueryParam alias', () => {
    expect(ODataParam).toEqual(ODataQueryParam)
  })

  test('OData param alternative constructor', () => {
    expect(OData.newParam()).toBeInstanceOf(ODataParam)
  })

  test('ODataParam skip and top', () => {
    const param = ODataParam.newParam().skip(30).top(10)
    expect(decodeURIComponent(param.toString())).toEqual("$skip=30&$top=10")
  })

  test('ODataParam orderby', () => {
    const param = ODataParam.newParam().orderby("CreationDateTime")
    expect(decodeURIComponent(param.toString())).toEqual("$orderby=CreationDateTime desc")
  })

  it('should support empty param', () => {
    const param = ODataParam.newParam()
    expect(param.toString()).toBe("")
  });

  test('ODataParam inlinecount', () => {
    const param = OData.newParam().inlinecount(true)
    expect(decodeURIComponent(param.toString())).toEqual("$inlinecount=allpages")
  })

  test('ODataParam filter', () => {
    const param = OData.newParam().filter(OData.newFilter().field("A").eqString("test"))
    expect(param.toString()).toEqual("$filter=A eq 'test'")
    const param2 = OData.newParam().filter("A eq 'test'")
    expect(decodeURIComponent(param2.toString())).toEqual("$filter=A eq 'test'")
    expect(() => {
      // @ts-ignore
      OData.newParam().filter(123)
    }).toThrow();
  })

  test('ODataParam inlinecount false', () => {
    const param = ODataParam.newParam().inlinecount(false)
    expect(decodeURIComponent(param.toString())).toEqual("")
  })

  test('ODataParam count true', () => {
    const param = ODataParam.newParam().count(true)
    expect(decodeURIComponent(param.toString("v4"))).toEqual("$count=true")
  })

  test('ODataParam orderby multi', () => {
    const param = ODataParam.newParam().orderbyMulti([{ field: "A" }, { field: "B", order: "asc" }]);
    expect(decodeURIComponent(param.toString())).toEqual("$orderby=A desc,B asc")
    const p2 = ODataParam.newParam().orderby([{ field: "A" }, { field: "B", order: "asc" }]);
    expect(decodeURIComponent(p2.toString())).toEqual("$orderby=A desc,B asc")
  })

  test('ODataParam format', () => {
    expect(() => {
      ODataParam.newParam().format("xml")
    }).toThrow();
    expect(decodeURIComponent(ODataParam.newParam().format("json").toString())).toEqual("$format=json")

  })

  test('ODataParam search', () => {
    const param = ODataParam.newParam().search("any word");
    expect(param.toString()).toEqual("$search=%any word%")
    const param2 = ODataParam.newParam().search("any word", false);
    expect(param2.toString()).toEqual("$search=any word")
  })

  test('ODataParam select', () => {
    const param = ODataParam.newParam().select("ObjectID");
    expect(decodeURIComponent(param.toString())).toEqual("$select=ObjectID")
  })

  test('ODataParam select (duplicate)', () => {
    const param = ODataParam.newParam().select(["ObjectID", "F1", "F1"]);
    expect(decodeURIComponent(param.toString())).toEqual("$select=ObjectID,F1")
  })

  test('ODataParam select multi', () => {
    const param = ODataParam
      .newParam()
      .select("ObjectID")
      .select("Name");
    expect(decodeURIComponent(param.toString())).toEqual("$select=ObjectID,Name")
  })

  test('expand navigation & replace', () => {
    const param = ODataParam.newParam().expand("Customer")
    expect(decodeURIComponent(param.toString())).toEqual("$expand=Customer")
    param.expand(["Customer", "Employee"], true)
    expect(decodeURIComponent(param.toString())).toEqual("$expand=Customer,Employee")
    expect(decodeURIComponent(ODataParam.newParam().expand("*").toString())).toEqual("$expand=*")
  })

  it('should support toString by varient', () => {
    expect(OData.newParam().custom("search", "v1").toString()).toBe("search=v1")
  });

  it('should support custom properties', () => {
    expect(OData.newParam().search("v1").toString("v2")).toBe("$search=%v1%")
    expect(OData.newParam().search("v1").toString("v2", "sap-gateway")).toBe("search=%v1%")
  });

  it('should support param.filter chain', () => {
    expect(
      OData.newParam()
        .filter()
        .property("a").eq(1)
        .property("value").eq(RawString.from('123'))
        .filterEnd().toString()
    ).toBe("$filter=a eq 1 and value eq 123")

  });


  it('should support ODataValueObject.toString', () => {
    expect(ODataDateTime.from(new Date("1995-11-11T00:00:00.000Z"), false).toString())
      .toBe("datetime'1995-11-11T00:00:00'")
    expect(ODataDateTime.from(new Date("1995-11-11T00:00:00.000Z"), true).toString())
      .toBe("datetime'1995-11-11T00%3A00%3A00'")
    expect(ODataDateTimeOffset.from(new Date("1995-11-11T00:00:00.000Z"), false).toString())
      .toBe("datetimeoffset'1995-11-11T00:00:00.000Z'")
    expect(ODataDateTimeOffset.from(new Date("1995-11-11T00:00:00.000Z"), true).toString())
      .toBe("datetimeoffset'1995-11-11T00%3A00%3A00.000Z'")
    expect(RawString.from("fdsajk@ðŸ˜‹2141'fdsaf").toString())
      .toBe("fdsajk@ðŸ˜‹2141'fdsaf")
  });

  it('should support $apply multi transformation with sequence', () => {
    expect(SystemQueryOptions.newOptions().apply(
      [
        transformation().groupBy(
          ['Time'],
          transformation().aggregate('Amount with sum as Total')
        ),
        transformation().aggregate('Total with average as DailyAverage'),
      ]
    ).toString('v4')).toMatchSnapshot()

  });

})
