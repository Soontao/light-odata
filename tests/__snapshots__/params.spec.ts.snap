// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ODataParams Test should support $apply multi transformation with sequence 1`] = `"$apply=groupBy((Time),aggregate(Amount with sum as Total))/aggregate(Total with average as DailyAverage)"`;
